resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ResourceTypeHelper.scala

```diff
@@ -0,0 +1,180 @@
+ * so we need to use reflection to avoid compile error when building against Hadoop 2.x
+ */
+object ResourceTypeHelper extends Logging {
+  private val AMOUNT_AND_UNIT_REGEX = "([0-9]+)([A-Za-z]*)".r
+  private val resourceTypesNotAvailableErrorMessage =
+    "Ignoring updating resource with resource types because " +
+    "the version of YARN does not support it!"
+
+  def setResourceInfoFromResourceTypes(
+      resourceTypesParam: Map[String, String],
+      resource: Resource): Resource = {
+    if (resource == null) {
+      throw new IllegalArgumentException("Resource parameter should not be null!")
+    }
+
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable()) {
+      logWarning(resourceTypesNotAvailableErrorMessage)
+      return resource
+    }
+
+    val resourceTypes = resourceTypesParam.map { case (k, v) => (
+      if (k.equals("memory")) {
+        logWarning("Trying to use 'memory' as a custom resource, converted it to 'memory-mb'")
+        "memory-mb"
+      } else k, v)
+    }
+
+    logDebug(s"Custom resource types: $resourceTypes")
+    resourceTypes.foreach { rt =>
+      val resourceName: String = rt._1
+      val (amount, unit) = getAmountAndUnit(rt._2)
+      logDebug(s"Registering resource with name: $resourceName, amount: $amount, unit: $unit")
+
+      try {
+        val resInfoClass = Utils.classForName(
+          "org.apache.hadoop.yarn.api.records.ResourceInformation")
+        val setResourceInformationMethod =
+          resource.getClass.getMethod("setResourceInformation", classOf[String],
+            resInfoClass)
+
+        val resourceInformation =
+          createResourceInformation(resourceName, amount, unit, resInfoClass)
+        setResourceInformationMethod.invoke(resource, resourceName, resourceInformation)
+      } catch {
+        case e: InvocationTargetException =>
+          if (e.getCause != null) {
+            throw e.getCause
+          } else {
+            throw e
+          }
+        case NonFatal(e) =>
+          logWarning(resourceTypesNotAvailableErrorMessage, e)
+      }
+    }
+    resource
+  }
+
+  def getCustomResourcesAsStrings(resource: Resource): String = {
+    if (resource == null) {
+      throw new IllegalArgumentException("Resource parameter should not be null!")
+    }
+
+    if (!ResourceTypeHelper.isYarnResourceTypesAvailable()) {
+      logWarning(resourceTypesNotAvailableErrorMessage)
+      return ""
+    }
+
+    var res: String = ""
+    try {
+      val resUtilsClass = Utils.classForName(
+        "org.apache.hadoop.yarn.util.resource.ResourceUtils")
+      val getNumberOfResourceTypesMethod = resUtilsClass.getMethod("getNumberOfKnownResourceTypes")
+      val numberOfResourceTypes: Int = getNumberOfResourceTypesMethod.invoke(null).asInstanceOf[Int]
+      val resourceClass = Utils.classForName(
+        "org.apache.hadoop.yarn.api.records.Resource")
+
+      // skip memory and vcores (index 0 and 1)
+      for (i <- 2 until numberOfResourceTypes) {
```

Comment:
I can agree with that, using an unstable interface could cause problems in the future.
For better traceability and informing users, I kept this method for now.
Maybe @vanzin could please suggest which is the better way (keep or not to keep)?

Reply:
Would be great to get @vanzin's opinion on this before merging. Thank you!

Reply:
Can you just print the `amResources` map from the user instead of having this hacky code? I'm not sure the YARN view of this is that interesting for Spark users, who are more interested in what they requested when debugging.

Reply:
Sorry but I don't get what is 'amResources' in this context or what exactly you had in mind for this.
I modified the code to append the toString() representation of the Resource field, so that I could completely remove the ResourceTypesHelper.getCustomResourcesAsStrings method which makes that class more simple.